contracts/Tokenator.sol

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity =0.8.24;
+
+import {ERC20, ERC20Burnable} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+
+contract Tokenator is ERC20Burnable, Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 underlyingToken;
+    uint256 private constant PRECISION = 1e18;
+    uint256 public immutable commencement;
+    uint256 public immutable durationSeconds;
+
+    error ZeroAddress();
+    error CommencementPassed();
+    error CommencementNotPassed();
+    error CannotRescueUnderlyingToken();
+
+    event Deployed(
+        address deployer,
+        IERC20 underlyingToken,
+        uint256 commencement,
+        uint256 durationDays
+    );
+    event ClaimUnderlyingToken(
+        address indexed user,
+        address indexed to,
+        uint256 tokenatorAmount,
+        uint256 underlyingTokenAmount
+    );
+    event EmergencyWithdrawal(uint256 amount, address indexed to);
+    event TokensRescued(IERC20 token, uint256 amount, address indexed to);
+
+    /// @notice following the airdrop, the deployer is required to burn any surplus Tokenator
+    /// (because totalSupply() is used in calculations) and send underlying token to the contract.
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        uint256 totalSupply_,
+        IERC20 underlyingToken_,
+        uint256 commencement_,
+        uint256 durationDays_
+    ) ERC20(name_, symbol_) Ownable(msg.sender) {
+        if (commencement_ < block.timestamp) revert CommencementPassed();
+
+        underlyingToken = underlyingToken_;
+        commencement = commencement_;
+        durationSeconds = durationDays_ * 1 days;
+
+        _mint(msg.sender, totalSupply_);
+
+        emit Deployed(
+            msg.sender,
+            underlyingToken_,
+            commencement_,
+            durationDays_
+        );
+    }
+
+    // redeem underlying token from this contract by burning tokenator
+    function claimUnderlyingToken(
+        uint256 tokenatorAmount,
+        address to
+    ) external returns (uint256 underlyingTokenAmount) {
+        if (commencement >= block.timestamp) revert CommencementNotPassed();
+        if (to == address(0)) revert ZeroAddress();
+
+        underlyingTokenAmount = underlyingTokenForTokenator(tokenatorAmount);
+        _burn(msg.sender, tokenatorAmount);
+        underlyingToken.transfer(to, underlyingTokenAmount);
```

Comment:
let's use safeTransfer here is well, even though we know what underlyingToken is

Reply:
done